crates/rustic_core/src/backend/s3.rs

```diff
@@ -0,0 +1,406 @@
+            .map(|region| Region::Custom { region, endpoint })
+            .map_err(|e| s3::region::error::RegionError::Env { source: e }),
+        None => Region::from_default_env(),
+    }
+    .map_err(S3ErrorKind::RegionError)?;
+
+    Ok(region)
+}
+
+fn object_path(tpe: FileType, id: Option<&Id>) -> String {
+    match tpe {
+        FileType::Config => format!("{}", tpe.dirname()),
+        FileType::Pack => {
+            if let Some(id) = id {
+                let hex_id = id.to_hex();
+                format!("{}/{}/{}", tpe.dirname(), &hex_id[..2], hex_id.to_string())
+            } else {
+                format!("{}/", tpe.dirname())
+            }
+        }
+        _ => {
+            if let Some(id) = id {
+                format!("{}/{}", tpe.dirname(), id.to_hex().to_string())
+            } else {
+                format!("{}/", tpe.dirname())
+            }
+        }
+    }
+}
+
+impl S3Backend {
+    pub fn new(url: &str) -> RusticResult<Self> {
+        let (endpoint, bucket) = parse_s3_url(url)?;
+
+        let region = get_s3_region(endpoint)?;
+        let credentials = Credentials::default().map_err(S3ErrorKind::CredentialsError)?;
+
+        let bucket = Bucket::new(&bucket, region, credentials)
+            .map_err(S3ErrorKind::S3Error)?
+            .with_path_style();
+
+        let backoff = WaitForLockBackoff::default();
+
+        Ok(Self { bucket, backoff })
+    }
+
+    fn path_style_url(&self) -> String {
+        format!(
+            "{}://{}/{}",
+            self.bucket.scheme(),
+            self.bucket.path_style_host(),
+            self.bucket.name()
+        )
+    }
+
+    fn call_with_retry(
+        &self,
+        f: impl Fn() -> Result<ResponseData, S3Error> + Send + Sync,
+    ) -> RusticResult<ResponseData> {
+        let backoff = self.backoff.clone();
+
+        let retry = || {
+            let response = f().map_err(|err| match err {
+                S3Error::WLCredentials => backoff::Error::Transient {
+                    err,
+                    retry_after: None,
+                },
+                S3Error::RLCredentials => backoff::Error::Transient {
+                    err,
+                    retry_after: None,
+                },
+                _ => {
+                    warn!("call_with_retry failed: {:?}", err);
+                    backoff::Error::Permanent(err)
+                }
+            })?;
+
+            Ok(response)
+        };
+
+        Ok(backoff::retry_notify(backoff, retry, notify)
+            .map_err(S3ErrorKind::BackoffError)?
+            .check_error()
+            .map_err(S3ErrorKind::S3Error)?)
+    }
+}
+
+impl ReadBackend for S3Backend {
+    fn location(&self) -> String {
+        let mut location = "s3:".to_string();
+        location.push_str(&self.path_style_url());
+        location
+    }
+
+    fn set_option(&mut self, _option: &str, _value: &str) -> RusticResult<()> {
+        todo!()
+    }
```

Comment:
I'm not sure what should be implemented here, so some help on this would be really appreciated